### ISSUE TITLE
Any possibility of a license change for sam.c/sam.h?

### ISSUE THREAD

**Issue by starseeker** [martanne/vis#238]
I suppose it's a long shot, but given how liberally licensed most of vis is (and how unusual that is in the world of text editors) has there been any attempt made to contact Lucent and see if they would be willing to re-license the sam.c file under ISC or something similar?

**Comment by martanne** [comment]
In general I would also prefer having the ~500 lines or so coming from sam licensed under ISC or MIT, if only for the fact that one less license would be involved.

The code under question can be found in both sam and acme:

* sam / acme
 - parse.h / edit.h
 - cmd.c / edit.c (mostly functions `parsecmd`, `simpleaddr`, `compoundaddr`)
 - xec.c / ecmd.c (`cmdexec`)
 - address.c / addr.c (`address`)

A first step would thus be to research under which license(s) these code bases are already available:

* the official [Plan 9 repository](https://9p.io/magic/webls?dir=/sources/plan9/) is licensed under [LPL 1.02](https://9p.io/sources/plan9/LICENSE) contains both [sam](https://9p.io/magic/webls?dir=/sources/plan9/sys/src/cmd/sam) and [acme](https://9p.io/magic/webls?dir=/sources/plan9/sys/src/cmd/acme)

* [Plan 9 from User Space](https://swtch.com/plan9port/) now [hosted on github](https://github.com/9fans/plan9port) is also [licensed under  LPL 1.02](https://github.com/9fans/plan9port/blob/master/LICENSE) but some parts of it, namely: `libutf`, `libfmt`, and `libregexp` are distributed under "simpler BSD-like boilerplates".

* [9base](http://tools.suckless.org/9base) contains sam and is also licensed under [LPL 1.02](http://git.suckless.org/9base/tree/LICENSE)

* In February 2014, the University of California, Berkeley, has been authorized by the current Plan 9 copyright holder – Alcatel-Lucent – to release all Plan 9 software previously governed by the Lucent Public License, Version 1.02 under the [GNU General Public License, Version 2](http://akaros.cs.berkeley.edu/files/Plan9License). The [source tarball](http://akaros.cs.berkeley.edu/files/plan9.tar.bz2) as well as [git repository](https://github.com/brho/plan9) contains both sam and acme. Hence the code is also available under GPLv2, probably not what you are looking for.

* Then there is [an updated version of sam](http://www.deadpixi.com/an-updated-version-of-sam) with accompaning [repository on github](https://github.com/deadpixi/sam) which claims to be "based on the original (?) port to X from 1989". According to the [LICENSE file](https://github.com/deadpixi/sam/blob/master/LICENSE) it is licensed under an ISC like license:

```
/*
 * The authors of this software are Rob Pike and Howard Trickey.
 * Copyright (c) 1998 by Lucent Technologies.
 *
 * Rob King made some changes.
 * Copyright (c) 2015 by Rob King.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */
```

* A file named `sam.shar` seems to be the [original Unix/X11 port of sam](http://www.netlib.org/research/sam.shar) referred to above. The corresponding [directory listing](http://www.netlib.org/research/) containing:
```
file  sam.shar
for   the Sam editor
by    Rob Pike, now maintained by Bob Flandrena <bobf@research.bell-labs.com>
size  976 kB
```
declares
```
#	All software in this directory is Copyright Lucent Technologies.
#	Permission to use, copy, modify, and distribute this software for any
#	purpose without fee is hereby granted, provided that this entire notice
#	is included in all copies of any software which is or includes a copy
#	or modification of this software and in all copies of the supporting
#	documentation for such software.
#	THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
#	WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT MAKE ANY
#	REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
#	OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
```

Some people also seem to have imported the content of the `sam.shar` source code archive into [git repositories hosted on github](https://github.com/floren/sam).

The README contains the following history lesson:
```
Rob Pike designed and implemented the original Unix version of sam and the
current version ported from Plan 9.  Howard Trickey provided the X
version of the graphics library, libXg.  Matty Farrow and his colleagues
at the University of Sydney extended libXg to support Runes.  Boyd Roberts
supplied the external command interface and the shell scripts for
the 'B' command. Doug Gwyn contributed many useful ideas to the X
implementation of sam. James Clark provided the simulations of the V10
Unix Man macros at the beginning of each manual page.
```
Also both the README and a file named [boilerplate](https://github.com/floren/sam/blob/master/boilerplate) contains the following license:
```
 * The authors of this software are Rob Pike and Howard Trickey.
 *		Copyright (c) 1998 by Lucent Technologies.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
```
Individual source files contain the following header `/* Copyright (c) 1998 Lucent Technologies - All rights reserved. */`

So to conclude it seems like the sam code is in fact already available under an ISC-like license. The difference to standard ISC license used by vis seems to be the additional requirement to also include the above copyright notice in "all copies of the supporting documentation for such software".

It might still be a good idea to reach out to the Plan 9 community (or directly to the original authors) and ask whether the above analysis is correct.

**Comment by starseeker** [comment]
Impressive analysis!  If you like I'll be glad to try emailing various folks, but I'm afraid it'll amount to cold calling since I don't know any of those key players personally...

Thanks for all your work on this - it's an exciting effort!

**Comment by starseeker** [comment]
Marc,
I just remember this issue was still open - did you want someone to try and reach the folks involved, or were you OK with using the copies of the files from the 1998 sam.shar version?

**Comment by starseeker** [comment]
I see on diffing that the sam.c in vis actually looks quite different from the sam.shar version, so maybe it isn't practical to use the older version after all...

**Comment by martanne** [comment]
Your approaching this form the wrong side. Of course the current `sam.c` file as used by vis looks quite different, after all it has been heavily adapted to use the existing vis functionality.

You would have to compare the relevant code sections (see first bullet points of my [initial comment](https://github.com/martanne/vis/issues/238#issuecomment-207950800) above) of the sam.shar version to the "current" ones found in 9base or plan9port which were roughly the starting point of my work.

* `parse.h` is identical (except for the Lucent Technologies copyright notice)
* `cmd.c` the relevant functions `parsecmd` and `compoundaddr` are identical. The only difference in `simpleaddr` is concerned with the initialization of a variable.
* `xec.c` the only relevant function from this file `cmdexec` is the same except for a renamed structure member in the otherwise unrelated `File` struct.
* `address.c` for the only relevant function of this file `address` the same is the case.

To summarize, the code vis re-uses in an adapted form can be traced back to the original X11 port of sam which is available under an ISC-like license.

**Comment by starseeker** [comment]
Ah, my bad.  I'm more used to situations where it is necessary to "redo" work from an original starting point that is clearly not under the undesirable license in question - I didn't think to do that sort of test.

So coming at it from that basis, are you comfortable applying the ISC-like license to your sam.c version, or is there more work you would like done that I can perhaps help with?
